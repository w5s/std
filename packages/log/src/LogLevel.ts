import type { Int, Option } from '@w5s/core';
import { IntBounded } from '@w5s/core/dist/Int/IntBounded.js';
import { Callable } from '@w5s/core/dist/Callable.js';
import { LogLevel as LogLevelType } from './LogLevel/LogLevel.js';
import { LogLevelComparable } from './LogLevel/LogComparable.js';

export interface LogLevel extends LogLevelType {}

function of(name: string, levelValue: Int): LogLevel {
  return { _: 'LogLevel', name, value: levelValue };
}

const None = of('None', IntBounded.minValue);

function match<T>(matchers: [LogLevel, T][]): (anyLevel: LogLevel) => Option<T>;
function match<T>(matchers: [LogLevel, T][], defaultValue: T): (anyLevel: LogLevel) => T;
function match<T>(matchers: [LogLevel, T][], defaultValue?: T): (anyLevel: LogLevel) => Option<T> {
  const orderedMatchers = [...matchers];
  const level = 0;
  const returnValue = 1;
  const first: [LogLevel, Option<T>] = [None, defaultValue];

  return orderedMatchers.length === 0
    ? () => defaultValue
    : (anyLevel) =>
        orderedMatchers.reduce(
          (acc, matcher) =>
            LogLevelComparable.compare(matcher[level], acc[level]) > 0 && anyLevel.value >= matcher[level].value
              ? matcher
              : acc,
          first,
        )[returnValue];
}

/**
 * @example
 * @namespace
 */
export const LogLevel = Callable({
  [Callable.symbol]: of,
  ...LogLevelType,
  ...LogLevelComparable,
  /**
   * Construct a new `LogLevel`
   *
   * @example
   * ```ts
   * const level = LogLevel('UberCritical', 60);// { levelName: 'UberCritical', level: 60 }
   * ```
   * @category Constructor
   * @param name - the level string representation
   * @param value - the level value
   */
  of,

  /**
   * Critical log level (50)
   */
  Critical: of('Critical', 50 as Int),
  /**
   * Error log level (40)
   */
  Error: of('Error', 40 as Int),
  /**
   * Warning log level (30)
   */
  Warning: of('Warning', 30 as Int),
  /**
   * Info log level (20)
   */
  Info: of('Info', 20 as Int),
  /**
   * Debug log level (10)
   */
  Debug: of('Debug', 10 as Int),

  /**
   * Returns the string representation of a level
   *
   * @example
   * ```ts
   * const level = LogLevel('Foo', 1);
   * LogLevel.format(level);// 'Foo'
   * ```
   * @param level - the log level
   */
  format(level: LogLevel): string {
    return level.name;
  },

  /**
   * Returns the numerical representation of a level
   *
   * @example
   * ```ts
   * const level = LogLevel('Foo', 1);
   * LogLevel.value(level);// 1
   * ```
   * @param level - the log level
   */
  value(level: LogLevel): number {
    return level.value;
  },

  /**
   * Build a matching function `(anyLevelValue) => value` from a list of tuples `[level1, value1], [level2, value2], ...`
   *
   * @example
   * ```ts
   * const matcher = LogLevel.match([
   *  [LogLevel.Error, 'This is error'],
   *  [LogLevel.Warning, 'This is warning'],
   * ]);
   * matcher(LogLevel.Critical);// 'This is error'
   * matcher(LogLevel.Error);// 'This is error'
   * matcher(LogLevel.Warning);// 'This is warning'
   * matcher(LogLevel.Info);// undefined
   * ```
   * @param matchers
   */
  match,
});
