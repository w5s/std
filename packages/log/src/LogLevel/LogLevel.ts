import { $Object as TObject } from '@w5s/core/dist/Type/Object.js';
import { string } from '@w5s/core/dist/Type/string.js';
import { Int as TInt } from '@w5s/core/dist/Type/Int.js';
import { constant } from '@w5s/core/dist/Type/constant.js';
import type { Type } from '@w5s/core';

export const LogLevel = TObject({
  _: constant('LogLevel'),
  /**
   * Log level name
   */
  name: string,
  /**
   * Log level value (number). Often used to filter / sort logs
   */
  value: TInt,
});

export interface LogLevel extends Type.TypeOf<typeof LogLevel> {}
