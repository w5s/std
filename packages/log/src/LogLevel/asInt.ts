import type { Int } from '@w5s/core';
import type { LogLevel } from './LogLevel.js';

/**
 * Returns the numerical representation of a level
 *
 * @example
 * ```typescript
 * const level = LogLevel('Foo', 1);
 * LogLevel.asInt(level);// 1
 * ```
 * @param level - the log level
 */
export function asInt(level: LogLevel): Int {
  return level.value;
}
