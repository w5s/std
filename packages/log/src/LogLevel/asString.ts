import type { LogLevel } from './LogLevel.js';

/**
 * Returns the string representation of a level
 *
 * @example
 * ```typescript
 * const level = LogLevel('Foo', 1);
 * LogLevel.asString(level);// 'Foo'
 * ```
 * @param self - the log level
 */
export function asString(self: LogLevel): string {
  return self.name;
}
