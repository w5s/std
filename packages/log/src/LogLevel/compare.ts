import type { Ordering } from '@w5s/core';
import { compare as compareNumber } from '@w5s/num/dist/Number/compare.js';
import type { LogLevel } from '../LogLevel.js';

export function compare(left: LogLevel, right: LogLevel): Ordering {
  return compareNumber(left.value, right.value);
}
