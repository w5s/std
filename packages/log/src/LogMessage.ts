import type { Type } from '@w5s/core';
import { Callable } from '@w5s/core/dist/Callable.js';
import { Array as TArray } from '@w5s/core/dist/Type/Array.js';
import { constant } from '@w5s/core/dist/Type/constant.js';
import { Record } from '@w5s/core/dist/Type/Record.js';
import { string } from '@w5s/core/dist/Type/string.js';
import { union } from '@w5s/core/dist/Type/union.js';
import { create } from './LogMessage/create.js';
import { of } from './LogMessage/of.js';

/**
 * @namespace
 */
export const LogMessageItemRef = Callable({
  [Callable.symbol]: (value: string): LogMessageItemRef => ({ $ref: value }),
  ...Record(constant('$ref'), string),
});
export type LogMessageItemRef = Type.TypeOf<typeof LogMessageItemRef>;

/**
 * @namespace
 */
export const LogMessageItem = union(string, LogMessageItemRef);

export type LogMessageItem = Type.TypeOf<typeof LogMessageItem>;

/**
 * @namespace
 */
export const LogMessage = Callable({
  [Callable.symbol]: of,
  ...TArray(LogMessageItem),
  create,
  of,
});

export interface LogMessage extends Type.TypeOf<typeof LogMessage> {}
