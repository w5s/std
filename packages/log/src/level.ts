import type { Option } from '@w5s/core';

export interface LogLevel {
  /**
   * The log level string representation
   */
  readonly logLevelName: string;
  /**
   * The log level numeric value
   */
  readonly logLevel: number;
}

/**
 * Construct a new `LogLevel`
 *
 * @example
 * ```ts
 * const level = LogLevel('UberCritical', 60);// { levelName: 'UberCritical', level: 60 }
 * ```
 * @category Constructor
 * @param levelName - the level string representation
 * @param level - the level value
 */
export function LogLevel(levelName: string, level: number): LogLevel {
  return { logLevelName: levelName, logLevel: level };
}
export namespace LogLevel {
  /**
   * Critical log level (50)
   */
  export const Critical = LogLevel('Critical', 50);
  /**
   * Error log level (40)
   */
  export const Error = LogLevel('Error', 40);
  /**
   * Warning log level (30)
   */
  export const Warning = LogLevel('Warning', 30);
  /**
   * Info log level (20)
   */
  export const Info = LogLevel('Info', 20);
  /**
   * Debug log level (10)
   */
  export const Debug = LogLevel('Debug', 10);

  const None = LogLevel('None', Number.NEGATIVE_INFINITY);

  /**
   * Return the comparison result
   *
   * @example
   * ```ts
   * const levels = [LogLevel('Lvl2', 2), LogLevel('Lvl1', 1), LogLevel('Lvl3', 3)];
   * levels.sort(Level.compare);// [LogLevel('Lvl1', 1), LogLevel('Lvl2', 2), LogLevel('Lvl3', 3)]
   * ```
   * @param left - the left element
   * @param right - the right element
   */
  export function compare(left: LogLevel, right: LogLevel): number {
    const diff = value(left) - value(right);

    return diff === 0 ? diff : diff < 0 ? -1 : 1;
  }

  /**
   * Returns the string representation of a level
   *
   * @example
   * ```ts
   * const level = LogLevel('Foo', 1);
   * LogLevel.stringify(level);// 'Foo'
   * ```
   * @param level - the log level
   */
  export function stringify(level: LogLevel): string {
    return level.logLevelName;
  }

  /**
   * Returns the numerical representation of a level
   *
   * @example
   * ```ts
   * const level = LogLevel('Foo', 1);
   * LogLevel.value(level);// 1
   * ```
   * @param level - the log level
   */
  export function value(level: LogLevel): number {
    return level.logLevel;
  }

  /**
   * Build a matching function `(anyLevelValue) => value` from a list of tuples `[level1, value1], [level2, value2], ...`
   *
   * @param matchers
   */
  export function match<T>(matchers: [LogLevel, T][]): (anyLevel: LogLevel) => Option<T>;
  export function match<T>(matchers: [LogLevel, T][], defaultValue: T): (anyLevel: LogLevel) => T;
  export function match<T>(matchers: [LogLevel, T][], defaultValue?: T): (anyLevel: LogLevel) => Option<T> {
    const orderedMatchers = [...matchers];
    const level = 0;
    const returnValue = 1;
    const first: [LogLevel, Option<T>] = [None, defaultValue];

    return orderedMatchers.length === 0
      ? () => defaultValue
      : (anyLevel) =>
          orderedMatchers.reduce(
            (acc, matcher) =>
              compare(matcher[level], acc[level]) > 0 && anyLevel.logLevel >= matcher[level].logLevel ? matcher : acc,
            first
          )[returnValue];
  }
}
