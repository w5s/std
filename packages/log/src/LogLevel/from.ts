import { LogLevel } from './LogLevel.js';
import { LogLevelValue } from './LogLevelValue.js';

const logLevelMap = Object.fromEntries(
  Object.entries(LogLevelValue).map(([k, v]) => [k.toLowerCase() as any, v]),
) as Record<LogLevelValue, LogLevel>;

/**
 *
 * @example
 * ```ts
 * LogLevel.from('critical') // LogLevel.Critical
 * ```
 * @param level
 */
export function from(value: LogLevelValue) {
  return logLevelMap[value];
}
