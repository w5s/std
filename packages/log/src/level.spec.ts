import { describe, it, expect } from 'vitest';
import { level } from './level.js';
import { LogLevel } from './LogLevel.js';
import { LogMessage, LogMessageRef } from './LogMessage.js';

describe(level, () => {
  it('returns a factory from level', () => {
    const fooValue = 'fooValue';
    const barValue = 'barValue';
    const message = level(LogLevel.Critical)`foo=${['foo', fooValue]},bar=${['bar', barValue]}`;

    expect(message).toEqual({
      level: LogLevel.Critical,
      message: LogMessage('foo=', LogMessageRef('foo'), ',bar=', LogMessageRef('bar')),
      data: {
        bar: barValue,
        foo: fooValue,
      },
    });
  });
  it('has .withData that adds more data', () => {
    const message = level(LogLevel.Critical).withData({ foo: 'foo_value' })`bar=${['bar', 'bar_value']}`;

    expect(message).toEqual({
      level: LogLevel.Critical,
      message: LogMessage('bar=', LogMessageRef('bar')),
      data: {
        bar: 'bar_value',
        foo: 'foo_value',
      },
    });
  });
  it('returns a factory from level value', () => {
    const fooValue = 'fooValue';
    const barValue = 'barValue';
    const message = level('critical')`foo=${['foo', fooValue]},bar=${['bar', barValue]}`;

    expect(message).toEqual({
      level: LogLevel.Critical,
      message: LogMessage('foo=', LogMessageRef('foo'), ',bar=', LogMessageRef('bar')),
      data: {
        bar: barValue,
        foo: fooValue,
      },
    });
  });
});
