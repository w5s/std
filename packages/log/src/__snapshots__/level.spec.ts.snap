// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogLevel > Critical > should have correct value > Critical 1`] = `
{
  "_": "LogLevel",
  "name": "Critical",
  "value": 50,
}
`;

exports[`LogLevel > Debug > should have correct value > Debug 1`] = `
{
  "_": "LogLevel",
  "name": "Debug",
  "value": 10,
}
`;

exports[`LogLevel > Error > should have correct value > Error 1`] = `
{
  "_": "LogLevel",
  "name": "Error",
  "value": 40,
}
`;

exports[`LogLevel > Info > should have correct value > Info 1`] = `
{
  "_": "LogLevel",
  "name": "Info",
  "value": 20,
}
`;

exports[`LogLevel > Warning > should have correct value > Warning 1`] = `
{
  "_": "LogLevel",
  "name": "Warning",
  "value": 30,
}
`;

exports[`LogLevel Critical should have correct value: Critical 1`] = `
{
  "_": "LogLevel",
  "name": "Critical",
  "value": 50,
}
`;

exports[`LogLevel Debug should have correct value: Debug 1`] = `
{
  "_": "LogLevel",
  "name": "Debug",
  "value": 10,
}
`;

exports[`LogLevel Error should have correct value: Error 1`] = `
{
  "_": "LogLevel",
  "name": "Error",
  "value": 40,
}
`;

exports[`LogLevel Info should have correct value: Info 1`] = `
{
  "_": "LogLevel",
  "name": "Info",
  "value": 20,
}
`;

exports[`LogLevel Warning should have correct value: Warning 1`] = `
{
  "_": "LogLevel",
  "name": "Warning",
  "value": 30,
}
`;
